rfd/0204-aws-iam-roles-anywhere.md

```diff
@@ -0,0 +1,769 @@
+
+#### End entity certificates - X.509
+When issuing X.509 certificates for users to access AWS, they must comply with the following:
+- must be a X.509v3 certificate
+- Basic Constraints must not include `CA` or its value must be false
+- the signing algorithm must include SHA256 or stronger
+- key usage must include Digital Signature
+
+[See more](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/trust-model.html#signature-verification)
+
+Sessions generated from IAM Roles Anywhere certificates have their [Source Identity](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_monitor.html) set to the common name (CN) of the certificate's Subject.
+Given its importance for auditing, this should contain the Teleport User name that requested the certificate.
+
+During IAM policy evaluation, the following certificate fields are exposed and can be used to allow or deny the request:
+- Subject fields as `PrincipalTag: aws:PrincipalTag/x509Subject/<field>` (eg CN maps to `PrincipalTag: aws:PrincipalTag/x509Subject/CN`)
+- Issuer fields as `PrincipalTag: aws:PrincipalTag/x509Issuer/<field>`
+- Subject Alternative Name (SAN) as `PrincipalTag: aws:PrincipalTag/x509SAN/<field>`
+
+Of the above, only some of them will be set by Teleport:
+- Subject Common Name: Teleport user that requested the access
+- Issuer Common Name: Teleport cluster name
+- Not After: this will be set based on the user's current session expiration
+
+### AWS Session creation
+When a client tool (Web UI or `tsh`) asks for a new AWS Session, Teleport Auth Service will generate the certificate, call `rolesanywhere.CreateSession` and exchange the certificate for AWS credentials.
+
+The call to [`rolesnaywhere.CreateSession`](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-create-session.html) is not explicit, but handled by the [rolesanywhere-credential-helper](https://github.com/aws/rolesanywhere-credential-helper) tool from AWS.
+
+The `rolesanywhere.CreateSession` call accepts a `durationSeconds` which indicates for how long the AWS Session will be valid.
+This durantion, per AWS documentation, cannot be higher than 12 hours, nor less than 15 minutes.
+
+When the Teleport User session is between 15 minutes and 12 hours, the AWS Session will have the same expiration.
+
+When the Teleport User session is lower than 15 minutes, the user will receive an error and is asked to login again.
+This is already the behavior on existing AWS Access flows. See [#46551](https://github.com/gravitational/teleport/issues/46551).
+
+When the Teleport User session is higher than 12 hours, the AWS Session will be at most 12 hours.
+
+### AWS Roles requirements for usage with IAM Roles Anywhere
+AWS IAM Roles must be accessible from Roles Anywhere service principal, which requires a [custom Trust Policy](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/trust-model.html#trust-policy).
```

Comment:
how will the user configure the trust policy for the IAM roles? do they have to repeat this manually?

Reply:
I have mixed feelings about this one.

If the user never used Roles Anywhere, the following steps are required:
- allow user to create Roles Anywhere Profiles from Teleport
- list all IAM Roles
- select and update the IAM Roles's trust policy
- assign IAM Role to the Profile
- sync Profiles to Teleport

My concern is how should the flow look like in order to use the `teleport integration configure` for the 3rd step?
Feels like we are helping them in a situation that:
- should be part of the AWS tooling/web console/UX
- it's only in the middle of the flow, they still need to do something manual afterwards (assign the IAM Role to the Profile)

Reply:
i thought `scripts/integration-setup-awsra.sh` was designed to create the profile. or does that just create the bootstrap role?

Reply:
It creates the necessary aws resources for the sync process.
It is not meant to be used by end-users for assuming IAM Roles.

It creates an IAM Role which can be used by the Trust Anchor (trust policy), and has permissions to list Roles Anywhere Profiles.
It also creates a Roles Anywhere Profile and adds the Role to its allowed roles.
This profile can only be used by the Auth Service and only for syncing Profiles into AWS Apps.